launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java

```diff
@@ -0,0 +1,335 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.launcher;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.apache.spark.launcher.CommandBuilderUtils.*;
+
+/**
+ * Special launcher for handling a CLI invocation of SparkSubmit.
+ * <p/>
+ * This launcher extends SparkLauncher to add command line parsing compatible with
+ * SparkSubmit. It handles setting driver-side options and special parsing needed
+ * for the different specialClasses.
+ * <p/>
+ * This class has also some special features to aid PySparkLauncher.
+ */
+class SparkSubmitCommandBuilder extends AbstractCommandBuilder {
+
+  /**
+   * Name of the app resource used to identify the PySpark shell. The command line parser expects
+   * the resource name to be the very first argument to spark-submit in this case.
+   *
+   * NOTE: this cannot be "pyspark-shell" since that identifies the PySpark shell to SparkSubmit
+   * (see java_gateway.py), and can cause this code to enter into an infinite loop.
+   */
+  static final String PYSPARK_SHELL = "pyspark-shell-main";
```

Comment:
it's just that right now I find the need to use an alternative name `pyspark-shell-main` a little difficult to reason about. Even as someone who is fairly familiar with how the pyspark shell is launched it took me a while to understand who is calling what, and why we might cause an infinite loop. IIUC the hierarchy is now as follows:
```
BASH scripts > Main.java > java_gateway.py > SparkSubmit.scala (another JVM)
```
whereas before we skipped the `Main.java` and went directly to python. AFAIK we're still launching a JVM as a python subprocess, so it doesn't simplify the subprocess hierarchy either.